lib/ffmpeg/qsv/encoder.py

```diff
@@ -59,6 +64,21 @@ def validate_caps(self):
 
     super().validate_caps()
 
+  def check_metrics(self):
+    # TCBRC check
+    if self.rcmode in ["vbr"] and vars(self).get("lowpower", 0) and vars(self).get("ldb", 0):
+      plimit = self.maxrate * 1000 / 8 / self.fps # Max allowed bytes per P-frame
+      output = call(
+        f"{exe2os('ffprobe')} -i {self.encoder.osencoded}"
+        f" -show_entries frame=pkt_size,pict_type -of compact",
+      )
+      frameszs = re.findall(r'(?<=pkt_size=).[0-9]*(?=\|pict_type=P)', output)
```

Comment:
if max_frame_size_i and max_frame_size_p is not defined, all frames should obey the average frame size. Not only P frames.

Reply:
In the cases I tested, I-frames were always too big and sometimes P/B frames are, too.  An "average" of all frame sizes might pass, though.

The check here is not "average"... it strictly checks **each** `frame size <= threshold`.  Should the check be `average(all frame sizes) <= threshold`, instead?

Reply:
I double check the VPL's code
1. Sorry I miss one option "-strict -1". To use TCBRC we need to disable NalHrdConformance. This is not listed in doc. I see this limitation in code [https://github.com/oneapi-src/oneVPL-intel-gpu/blob/4c41eeff76ab594575fe3b7519df478ee6d2fd7d/_studio/mfx_lib/encode_hw/hevc/agnostic/base/hevcehw_base_legacy.cpp#L293](https://github.com/oneapi-src/oneVPL-intel-gpu/blob/4c41eeff76ab594575fe3b7519df478ee6d2fd7d/_studio/mfx_lib/encode_hw/hevc/agnostic/base/hevcehw_base_legacy.cpp#L293 )
2. I think we should check `average(all frame) < threshold` and the target average frame size is calculated by "`targetkbps * 1000 / 8 / framerate`" .
3. low_delay_brc also automatically set max_frame_size limitation. max_frame_size is calculated based on maxKbps, but vary on different codec. For h264, it is (`maxkbps * 1000 / 8 / framerate`)[https://github.com/oneapi-src/oneVPL-intel-gpu/blob/4c41eeff76ab594575fe3b7519df478ee6d2fd7d/_studio/mfx_lib/encode_hw/h264/src/mfx_h264_enc_common_hw.cpp#L5204](https://github.com/oneapi-src/oneVPL-intel-gpu/blob/4c41eeff76ab594575fe3b7519df478ee6d2fd7d/_studio/mfx_lib/encode_hw/h264/src/mfx_h264_enc_common_hw.cpp#L5204). For hevc, it is (`1.4 * maxkbps * 1000 / 8 / framerate`) [https://github.com/oneapi-src/oneVPL-intel-gpu/blob/4c41eeff76ab594575fe3b7519df478ee6d2fd7d/_studio/mfx_lib/encode_hw/hevc/agnostic/base/hevcehw_base_max_frame_size.cpp#L102](https://github.com/oneapi-src/oneVPL-intel-gpu/blob/4c41eeff76ab594575fe3b7519df478ee6d2fd7d/_studio/mfx_lib/encode_hw/hevc/agnostic/base/hevcehw_base_max_frame_size.cpp#L102). To see the exact value, you can see ffmpeg log when run command with "-v verbose" option. To test this feature, I suggest to test "`(each frame size) < max_frame_size`" then check "`(failed frame number)/(total frame number) < threshold`". I don't know if we should check this as well, it seems relate to low_delay_brc but doesn't belong to TCBRC.

Reply:
Ok, I updated PR to 1. and 2.  I still get a lot of test cases fail the TCBRC check.  Please try on your side and advise.  You'll need to define some cases on your side as I have not created a patch for them.

Reply:
I tried on my side. many failed case as well. Most of them is caused by low bitrate. I will check with VPL team. Seems I misunderstand some TCBRC's behavior.

Reply:
I confirmed with VPL team.
1. "targetKbps*1000/8/fps" is a reference number. Driver try to make pkt_size as close to this number as possible. When VPL team test this feature they set a boundary like this "`each pkt_size < 1.2 * target_frame_size`". "`1.2`" is not a official number. It depends on video content. I think we can adjust it according to our CI test result.
2. Even we increase the boundary. There are some frame bigger than this number, for example I frame. VPL team said it is expected. We can set another check "`failed frame number/ total frame number < tolerance rate`".
3. According to VPL team, bitrate too low is probably a bug. I will submit a ticket to VPL team.

Reply:
Updated.  Please try now.

Reply:
except the bitrate problem, It works on my side.

Reply:
LGTM, Thanks a lot for you working😄. Sorry for bothering you to change so many times.